### PR TITLE
Implement rank ordinal as main ranking algorithm for NSGA-II

### PR DESCRIPTION
**Pull request by lukepmccombs**
Implements rank ordinal from [Rank-based Non-dominated Sorting](https://arxiv.org/abs/2203.13654) as a replacement to `fast_nondominated_sort`for NSGA-II's default sorting function.
output

Renamed `nsga2` to `generalized_nsga2`, adding a new `rank_func` parameter.

Added manually calculated ranks to `generate_test_pop()` for multiobjective operator tests.
Added rudimentary test for `rank_ordinal_sort`.

Created a second multiobjective / NSGA-II notebook that tests output and demonstrates the speedup.

### CODE REVIEW

File: CHANGELOG.md
Comment (SigmaX) [inline CHANGELOG.md:19]: Ope, were these two lines removed on accident?  Perhaps a merge mistake?

File: leap_ec/multiobjective/nsga2.py
Comment (SigmaX) [inline leap_ec/multiobjective/nsga2.py:31]: Let's add a sentence explaining what we mean by "generalized."  Something like:

```python
"""
NSGA-II multi-objective evolutionary algorithm.  This classic algorithm relies on the idea of "non-dominated sorting" and de-crowding to evolve a diverse Pareto front.  The "generalized" NSGA-II we implement here differs slightly from the canonical algorithm, in that we default to a faster sorting algorithm devised by Burlacu (2022).

- Deb, Kalyanmoy, Amrit Pratap, Sameer Agarwal, and T. A. M. T. Meyarivan.
          "A Fast and Elitist Multiobjective Genetic Algorithm: NSGA-II." IEEE
          transactions on evolutionary computation 6, no. 2 (2002): 182-197.

 - Bogdan Burlacu. 2022. Rank-based Non-dominated Sorting. arXiv.
          DOI:https://doi.org/10.48550/ARXIV.2203.13654 

If you wish the algorithm to use the original NSGA-II behavior instead (which runs much slower), you can select the original operator by passing in `rank_fun=fast_nondominated_sort`.
"""
```

File: leap_ec/multiobjective/nsga2.py
Comment (SigmaX) [inline leap_ec/multiobjective/nsga2.py:34]: It would be nice to have a doctest.  An example of what this looks like is in the docstring for `generational_ea` [here](https://github.com/AureumChaos/LEAP/blob/dba570f6c7153e5dc43e126815030612aee9058a/leap_ec/algorithm.py#L78).

Doctests for high-level algorithms like this are verbose, but nice because they mean a complete usage example is included in the docs.

File: tests/multiobjective/test_ops.py
Comment (SigmaX) [inline tests/multiobjective/test_ops.py:50]: This is great! I think it's good enough for an initial PR.

Wondering if we can think up one or two more more test inputs, though.  I'm not sure what the boundary conditions are that we should check, but I imagine there are a few kinds of bugs that this test would not catch. Maybe we can add a test using an additional problem?  Either a follow-on task for @markcoletti or something @lukepmccombs could add now if he has time.

**Review by SigmaX — CHANGES_REQUESTED** [review]
This is awesome, @lukepmccombs! Thanks for the contribution.

I took a quick pass and left some minor comments.  I didn't run it myself or look at the notebooks— @markcoletti may have additional comments on those.

One more thing: I think it would be great to go ahead and add an `nsga_ii()` wrapper that sets the ranking operator automatically.  We can easily do that ourselves later though, so optional for this PR.

File: CHANGELOG.md
Comment (lukepmccombs) [reply CHANGELOG.md:19]: [This change](https://github.com/AureumChaos/LEAP/pull/223/commits/f16306357a54e12eed71df5813e37899600ecbeb) was made by @markcoletti after he merged NSGA-II back into develop, where I branched off.

File: leap_ec/multiobjective/nsga2.py
Comment (lukepmccombs) [reply leap_ec/multiobjective/nsga2.py:31]: I added your comment into the docstring and formatted it to match the rest. I didn't have anything else to add to it.

File: leap_ec/multiobjective/nsga2.py
Comment (lukepmccombs) [reply leap_ec/multiobjective/nsga2.py:70]: I wrote one, its fairly rudimentary but it serves as a usage example. I used `random_selection` rather than `tournament_selection` in the example to match the original paper.

**Review by SigmaX — APPROVED** [review]
Looks good! Thanks for the changes, @lukepmccombs — appreciate it.

A few notebook tests are failing that are not touched by this PR, but that's an independent issue.  Besides that tests pass.

There are a couple things we may want to polish—the narrative for the comparison notebook, for instance, and may want to add `test_env` environment variable logic to notebooks so they run quick in test mode.  But IMO that shouldn't hold up this PR.

Leaving it to @markcoletti for the last word.

**Comment by markcoletti** [comment]
> Looks good! Thanks for the changes, @lukepmccombs — appreciate it.
> 
> A few notebook tests are failing that are not touched by this PR, but that's an independent issue. Besides that tests pass.
> 
> There are a couple things we may want to polish—the narrative for the comparison notebook, for instance, and may want to add `test_env` environment variable logic to notebooks so they run quick in test mode. But IMO that shouldn't hold up this PR.
> 
> Leaving it to @markcoletti for the last word.

I've looked through the code and it looks fine by me.  Thanks @lukepmccombs for this work!  Merging in pull request now.

**Pull request merged**
